File: 2022/day1/day1.ts
const input_text : string = await Deno.readTextFile("day1.input")
const input_list : string[] = input_text.split("\n");

type Elf = {
	Calories: number[]
	CalSum: number
}

function parseInput(inputs :string[]) {
	let result = [] as Elf[]
	
	let currentElf = 0
	for (let i = 0; i < inputs.length; i++) {
		if (result[currentElf] == undefined) {
			result[currentElf] = {Calories: [], CalSum: 0}
		}
		if (inputs[i] == "") {
			currentElf++
			continue
		}
		result[currentElf].Calories.push(parseInt(inputs[i]))
		result[currentElf].CalSum += parseInt(inputs[i])
	}
	return result
}

function sortElfsByCalories(elfs :Elf[]) {
	return elfs.sort((a, b) => (a.CalSum < b.CalSum) ? 1 : -1)
}

const elfs = parseInput(input_list)
const sortedElfs = sortElfsByCalories(elfs)
console.log(sortedElfs.slice(0,3).reduce((a, b) => a +  b.CalSum, 0))

File: 2022/day1/day1.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Elf struct {
	Calories []int
	CalSum   int
}

func PanicIfErr(err error) {
	if err != nil {
		panic(err)
	}
}

func parseInput() []Elf {
	byteArr, err := os.ReadFile("day1.input")
	PanicIfErr(err)
	inputstr := string(byteArr)
	inputarr := strings.Split(inputstr, "\n")

	elfs := []Elf{}
	currentelf := 0

	for _, v := range inputarr {
		if len(elfs) <= currentelf {
			elfs = append(elfs, Elf{Calories: nil, CalSum: 0})
		}

		if v == "" {
			currentelf++
			continue
		}

		intval, err := strconv.Atoi(v)
		PanicIfErr(err)
		elfs[currentelf].Calories = append(elfs[currentelf].Calories, intval)
		elfs[currentelf].CalSum += intval
	}

	return elfs
}

func sortElfsByCalories(elfs []Elf) {
	sort.Slice(elfs, func(i, j int) bool {
		return elfs[i].CalSum > elfs[j].CalSum
	})
}

func main() {
	elfs := parseInput()
	sortElfsByCalories(elfs)
	fmt.Printf("%v", elfs[0].CalSum+elfs[1].CalSum+elfs[2].CalSum)
}
